File: sdk/java/src/main/java/internal/telemetry/Telemetry.java
package internal.telemetry;

import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;

/**
 * The process-wide client, and a facade over it.
 *
 * <p>Application code reads as one line at the point of interest, with no plumbing:
 * <pre>{@code
 * Telemetry.signal("console.approvals.decision.role_gate_missed",
 *                  SignalOptions.payload(decision).withDetail("recorded by " + role));
 * }</pre>
 *
 * <p>Construction is lazy: code that records something before — or without — an
 * explicit {@link #init()} still records, rather than raising. With no
 * {@code TELEMETRY_SERVICE} in the environment the client is inert and every method
 * here is a no-op, so a service runs unchanged on a workstation with no collector.
 */
public final class Telemetry {

    private static volatile TelemetryClient active;
    private static final Object LOCK = new Object();

    private Telemetry() {
    }

    /** Create, or replace, the process-wide client from the environment. */
    public static TelemetryClient init() {
        return init(TelemetryConfig.fromEnvironment());
    }

    public static TelemetryClient init(TelemetryConfig config) {
        TelemetryClient replaced;
        TelemetryClient created = new TelemetryClient(config);
        synchronized (LOCK) {
            replaced = active;
            active = created;
        }
        if (replaced != null) {
            // Draining the old one must not make the caller wait on a collector.
            Thread.ofVirtual().start(() -> {
                replaced.flush(Duration.ofMillis(500));
                replaced.close(Duration.ofMillis(500));
            });
        }
        return created;
    }

    /** The process-wide client, built from the environment on first use. */
    public static TelemetryClient get() {
        TelemetryClient current = active;
        if (current == null) {
            synchronized (LOCK) {
                if (active == null) {
                    active = new TelemetryClient();
                }
                current = active;
            }
        }
        return current;
    }

    public static void signal(String name) {
        get().signal(name);
    }

    public static void signal(String name, SignalOptions options) {
        get().signal(name, options);
    }

    public static void signal(String name, Map<String, Object> attributes) {
        get().signal(name, attributes);
    }

    public static void note(String message) {
        get().note(message);
    }

    public static String outbound(String destination, EgressDeclaration declaration) {
        return get().outbound(destination, declaration);
    }

    public static String outbound(String destination, String signal, String param) {
        return get().outbound(destination, signal, param);
    }

    public static void contribute(List<Attribute> attributes, String route, String method) {
        get().contribute(attributes, route, method);
    }

    public static void authSubject(String subject) {
        get().authSubject(subject);
    }

    public static String currentRequestId() {
        return get().currentRequestId();
    }

    /** Carry the in-flight request's facts into work handed to a bare executor. */
    public static Runnable wrap(Runnable task) {
        return TelemetryContext.wrap(task);
    }

    public static <T> Callable<T> wrap(Callable<T> task) {
        return TelemetryContext.wrap(task);
    }

    public static ExecutorService propagating(ExecutorService delegate) {
        return TelemetryContext.propagating(delegate);
    }

    public static boolean flush(Duration budget) {
        return get().flush(budget);
    }

    public static Map<String, Long> stats() {
        return get().stats();
    }

    /** For tests: drop the process-wide client so the next call builds a fresh one. */
    static void reset() {
        TelemetryClient previous;
        synchronized (LOCK) {
            previous = active;
            active = null;
        }
        if (previous != null) {
            previous.close(Duration.ofMillis(200));
        }
    }
}

File: sdk/dotnet/Internal.Telemetry/Telemetry.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Internal.Telemetry;

/// <summary>
/// The process-wide client, and a one-line facade over it.
/// </summary>
/// <remarks>
/// Application code reads as <c>Telemetry.Signal("portal.documents.write.path_escape",
/// detail)</c>: one line at the point of interest, no plumbing. Services that prefer
/// dependency injection resolve <see cref="TelemetryClient"/> instead; both names refer
/// to the same instance once the hosting extensions have run.
/// </remarks>
public static class Telemetry
{
    private static TelemetryClient? _current;
    private static readonly object Gate = new();

    /// <summary>
    /// The active client, built from the environment on first use.
    /// </summary>
    /// <remarks>
    /// Lazy construction matters: code that raises a counter before, or without, an
    /// explicit <see cref="Init"/> must still record rather than throw.
    /// </remarks>
    public static TelemetryClient Current
    {
        get
        {
            TelemetryClient? client = Volatile.Read(ref _current);
            if (client is not null)
            {
                return client;
            }

            lock (Gate)
            {
                _current ??= new TelemetryClient();
                return _current;
            }
        }
    }

    /// <summary>Create, or replace, the process-wide client.</summary>
    /// <param name="options">Explicit options, or null to read the environment alone.</param>
    /// <returns>The new client.</returns>
    public static TelemetryClient Init(TelemetryOptions? options = null)
    {
        TelemetryClient client = new(options);
        return Install(client);
    }

    /// <summary>Adopt an already-built client as the process-wide one.</summary>
    /// <param name="client">The client to adopt.</param>
    /// <returns>The adopted client.</returns>
    public static TelemetryClient Install(TelemetryClient client)
    {
        if (client is null)
        {
            throw new ArgumentNullException(nameof(client));
        }

        TelemetryClient? previous;
        lock (Gate)
        {
            previous = _current;
            _current = client;
        }

        if (previous is not null && !ReferenceEquals(previous, client))
        {
            // Fire and forget: replacing the client must not make a caller wait on a
            // drain of the old one.
            _ = previous.DisposeAsync().AsTask();
        }

        return client;
    }

    /// <summary>Raise a counter on the process-wide client.</summary>
    /// <param name="name">Dotted metric name.</param>
    /// <param name="payload">The input that produced the anomaly.</param>
    /// <param name="detail">What was actually observed.</param>
    /// <param name="requestId">Correlation id, when the caller tracks one.</param>
    public static void Signal(string name, string? payload = null, string? detail = null, string? requestId = null)
    {
        Current.Signal(name, payload, detail, requestId);
    }

    /// <summary>Write a breadcrumb on the process-wide client.</summary>
    /// <param name="message">The message.</param>
    public static void Note(string message) => Current.Note(message);

    /// <summary>
    /// Carry the facts of the request in flight into work that runs somewhere the
    /// execution context does not reach. See <see cref="TelemetryContext.Bind(Action)"/>.
    /// </summary>
    /// <param name="body">The work to carry the context into.</param>
    /// <returns>A wrapper that puts the current context in scope while it runs.</returns>
    public static Action Bind(Action body) => TelemetryContext.Bind(body);

    /// <summary>Carry the facts of the request in flight into asynchronous work.</summary>
    /// <param name="body">The work to carry the context into.</param>
    /// <returns>A wrapper that puts the current context in scope while it runs.</returns>
    public static Func<Task> Bind(Func<Task> body) => TelemetryContext.Bind(body);

    /// <summary>Register an outbound dependency call on the process-wide client.</summary>
    /// <param name="destination">The URL or host about to be resolved.</param>
    /// <param name="signal">The code path making the call.</param>
    /// <param name="param">Name of the input the destination came from.</param>
    /// <param name="route">Route template; taken from the request when omitted.</param>
    /// <returns>The correlation id.</returns>
    public static string Outbound(string destination, string signal, string? param = null, string? route = null)
    {
        return Current.Outbound(destination, signal, param, route);
    }
}
